core/include/io.h

```diff
@@ -273,4 +273,33 @@ static inline void io_clrsetbits8(vaddr_t addr, uint8_t clear_mask,
 	io_write8(addr, (io_read8(addr) & ~clear_mask) | set_mask);
 }
 
+/*
+ * The correct value was read or the ops timed out.
+ *
+ * @addr is the address of the memory cell accessed
+ * @val represents the val  of the memory cell accessed
+ * @cond represents the condition to get the correct value
+ * @delay_us represents the read interval
+ * @timeout_us represents the time-out period
+ *
+ * @return true means ops timed out, false means got right value
+ */
+#define IO_READ32_POLL_TIMEOUT(addr, val, cond, delay_us, timeout_us) \
+({ \
+	uint64_t _timeout = 0; \
+	uint64_t _delay_us = delay_us; \
+	bool _flag = true; \
+	\
+	while (_timeout < (timeout_us)) { \
+		(val) = io_read32(addr); \
+		if (cond) { \
+			_flag = false; \
+			break; \
+		} \
+		_timeout += (_delay_us); \
+		udelay(_delay_us); \
+	} \
+	(_flag); \
```

Comment:
If the thread from where `IO_READ32_POLL_TIMEOUT()` is called is rescedule by non-secure world, we may timeout while the target IO memory cell is at the expected state. Could you add a final test on `cond` test once out of the loop? Something like:
```patch
 		_timeout += (_delay_us); \
 		udelay(_delay_us); \
 	} \
+	(val) = io_read32(addr); \
+	if (cond) { \
+		_flag = false; \
 	(_flag); \
```

Note: can remove the parentheses around `flags` in the last line.